File: parser/parser.go
package parser

import (
	"errors"
	"fmt"
	"strconv"

	"github.com/juancortelezzi/alambre/ast"
	"github.com/juancortelezzi/alambre/token"
)

var (
	ErrInvalidConversion = errors.New("Parser: invalid conversion")
	ErrInvalidToken      = errors.New("Parser: invalid token")
	ErrUnknownIdentifier = errors.New("Parser: unknown identifier")
	// input := `"\n" split (chars (to_int) map (is_some) filter first unwrap 10 * swap last unwrap swap drop +) map 0 (+) reduce`
	builtins = map[string]struct{}{
		"reduce":  {},
		"map":     {},
		"status":  {},
		"split":   {},
		"chars":   {},
		"to_int":  {},
		"filter":  {},
		"is_some": {},
		"first":   {},
		"unwrap":  {},
		"swap":    {},
		"last":    {},
		"drop":    {},
	}
)

type Parser struct {
	program   []ast.Node
	currToken token.Token
	peekToken token.Token
	nextToken func() token.Token
}

func NewParser(nextToken func() token.Token) *Parser {
	currToken := nextToken()
	peekToken := nextToken()
	return &Parser{
		program:   make([]ast.Node, 0, 10),
		currToken: currToken,
		peekToken: peekToken,
		nextToken: nextToken,
	}
}

func NewParserFromTokens(tokens []token.Token) *Parser {
	index := 0
	nextToken := func() token.Token {
		if index >= len(tokens) {
			return token.Token{Type: token.EOF, Literal: ""}
		}
		tok := tokens[index]
		index += 1
		return tok
	}

	return NewParser(nextToken)
}

func (p *Parser) readToken() {
	p.currToken = p.peekToken
	p.peekToken = p.nextToken()
}

func (p *Parser) parseNumber() (ast.Node, error) {
	tok := p.currToken
	value, err := strconv.ParseFloat(tok.Literal, 64)
	if err != nil {
		return nil, fmt.Errorf(
			"%w: converting '%s' as number at pos '%d:%d'",
			ErrInvalidConversion,
			tok.Literal,
			tok.Row,
			tok.Col,
		)
	}

	p.readToken()

	return &ast.Number{Value: value, Col: tok.Col, Row: tok.Row}, nil
}

func (p *Parser) parseString() ast.Node {
	tok := p.currToken
	p.readToken()
	return &ast.String{Value: tok.Literal, Col: tok.Col, Row: tok.Row}
}

func (p *Parser) parseBinop() ast.Node {
	tok := p.currToken
	var binopType ast.BinopType
	switch tok.Type {
	case token.Plus:
		binopType = ast.Addition
	case token.Minus:
		binopType = ast.Subtraction
	case token.Asterisk:
		binopType = ast.Multiplication
	case token.Slash:
		binopType = ast.Division
	case token.LessThan:
		binopType = ast.LessThan
	case token.LessThanEqual:
		binopType = ast.LessThanEqual
	case token.GreaterThan:
		binopType = ast.GreaterThan
	case token.GreaterThanEqual:
		binopType = ast.GreaterThanEqual
	case token.Equal:
		binopType = ast.Equal
	case token.NotEqual:
		binopType = ast.NotEqual
	case token.And:
		binopType = ast.And
	case token.Or:
		binopType = ast.Or
	case token.Not:
		binopType = ast.Not
	}

	p.readToken()
	return &ast.Binop{Type: binopType, Col: tok.Col, Row: tok.Row}
}

func (p *Parser) parseTable() (ast.Node, error) {
	if p.currToken.Type != token.LSquiggly {
		// NOTE: this is a bad error, this should not happen if I am a good
		// programmer. Too lazy to change it though.
		return nil, fmt.Errorf("%w: expected '{' at pos '%d:%d'", ErrInvalidToken, p.currToken.Row, p.currToken.Col)
	}

	col := p.currToken.Col
	row := p.currToken.Row

	p.readToken()

	table := &ast.Table{Items: make([]ast.TableItem, 0), Col: col, Row: row}
	tableItem := ast.TableItem{Value: make([]ast.Node, 0)}

	for {
		if len(tableItem.Value) == 0 {
			isStringOrIdent := p.currToken.Type == token.String || p.currToken.Type == token.Ident
			hasColon := p.peekToken.Type == token.Colon
			if isStringOrIdent && hasColon {
				tableItem.IsKvp = true
				tableItem.Key = p.currToken.Literal
				p.readToken()
				p.readToken()
				continue
			}
		}

		node, err := p.parseExpression()
		if err != nil && errors.Is(err, ErrInvalidToken) {
			if p.currToken.Type == token.Comma {
				table.Items = append(table.Items, tableItem)
				tableItem = ast.TableItem{Value: make([]ast.Node, 0)}
				p.readToken()
				continue
			}

			if p.currToken.Type == token.RSquiggly {
				table.Items = append(table.Items, tableItem)
				p.readToken()
				break
			}

			if p.currToken.Type == token.Colon {
				// p.readToken()
				panic("todo")
			}

			panic("this should not happen")
		}

		if err != nil {
			return nil, err
		}

		tableItem.Value = append(tableItem.Value, node)
	}

	return table, nil
}

func (p *Parser) parseIdent() (ast.Node, error) {
	tok := p.currToken
	p.readToken()

	if tok.Literal == "true" {
		return &ast.Boolean{Value: true, Col: tok.Col, Row: tok.Row}, nil
	}

	if tok.Literal == "false" {
		return &ast.Boolean{Value: false, Col: tok.Col, Row: tok.Row}, nil
	}

	if _, ok := builtins[tok.Literal]; ok {
		return &ast.FnCall{Name: tok.Literal, Col: tok.Col, Row: tok.Row}, nil
	}

	return &ast.Ident{Name: tok.Literal, Col: tok.Col, Row: tok.Row}, nil
}

func (p *Parser) parseFn() (ast.Node, error) {
	if p.currToken.Type != token.LParen {
		// NOTE: this is a bad error, this should not happen if I am a good
		// programmer. Too lazy to change it though. v2.
		return nil, fmt.Errorf("%w: expected '(' at pos '%d:%d'", ErrInvalidToken, p.currToken.Row, p.currToken.Col)
	}

	col := p.currToken.Col
	row := p.currToken.Row

	p.readToken()

	fn := &ast.Fn{Program: make([]ast.Node, 0), Col: col, Row: row}
	for {
		node, err := p.parseExpression()
		if err != nil && errors.Is(err, ErrInvalidToken) {
			if p.currToken.Type != token.RParen {
				panic("this should not happen")
			}
			p.readToken()
			return fn, nil
		}

		if err != nil {
			return nil, err
		}

		fn.Program = append(fn.Program, node)
	}
}

func (p *Parser) parseExpression() (ast.Node, error) {
	for {
		t := p.currToken

		switch t.Type {
		case token.Number:
			node, err := p.parseNumber()
			return node, err

		case token.String:
			node := p.parseString()
			return node, nil

		case token.LSquiggly:
			node, err := p.parseTable()
			return node, err

		case token.LParen:
			node, err := p.parseFn()
			return node, err

		case token.Ident:
			node, err := p.parseIdent()
			return node, err

		case token.Plus,
			token.Minus,
			token.Asterisk,
			token.Slash,
			token.LessThan,
			token.LessThanEqual,
			token.GreaterThan,
			token.GreaterThanEqual,
			token.Equal,
			token.NotEqual,
			token.And,
			token.Or,
			token.Not:
			node := p.parseBinop()
			return node, nil

		default:
			return nil, fmt.Errorf("%w: token '%s' at pos '%d:%d'", ErrInvalidToken, t.Literal, t.Row, t.Col)
		}
	}
}

func (p *Parser) Parse() ([]ast.Node, error) {
	for p.currToken.Type != token.EOF {
		node, err := p.parseExpression()
		if err != nil {
			return nil, err
		}
		p.program = append(p.program, node)
	}

	return p.program, nil
}

File: src/parser.zig
const std = @import("std");
const ast = @import("ast.zig");
const token = @import("tokens.zig");
const assert = std.debug.assert;

const ParserErrType = error{
    InvalidToken,
    UnknownIdentifier,
    InvalidConversion,
    NotEnoughArguments,
};

const ParseErr = struct {
    type: ParserErrType,
    row: usize,
    col: usize,
};

pub const Parser = struct {
    allocator: std.mem.Allocator,
    tokens: []const token.Token,
    program: std.ArrayList(ast.Node),
    indexes: std.ArrayList(u32),
    cur: token.Token,
    read_pos: usize,

    pub fn init(allocator: std.mem.Allocator, tokens: []const token.Token) Parser {
        assert(tokens.len > 0);
        return Parser{
            .allocator = allocator,
            .tokens = tokens,
            .program = std.ArrayList(ast.Node).init(allocator),
            .indexes = std.ArrayList(u32).init(allocator),
            .cur = tokens[0],
            .read_pos = 1,
        };
    }

    pub fn deinit(self: *Parser) void {
        self.program.deinit();
        self.indexes.deinit();
    }

    fn advance(self: *Parser) void {
        if (self.read_pos >= self.tokens.len) {
            return;
        }

        self.cur = self.tokens[self.read_pos];
        self.read_pos += 1;
    }

    fn parseNumber(self: *Parser) ?ParseErr {
        const tok = self.cur;
        const value = std.fmt.parseFloat(f64, tok.literal) catch {
            return ParseErr{
                .type = ParserErrType.InvalidConversion,
                .row = tok.row,
                .col = tok.col,
            };
        };

        self.advance();

        const number = ast.Node{ .number = .{ .val = value } };

        self.program.append(number) catch @panic("failed to append node");
        self.indexes.append(@intCast(self.program.items.len - 1)) catch @panic("failed to append index");
        return null;
    }

    fn parseString(self: *Parser) ?ParseErr {
        const tok = self.cur;
        self.advance();
        const string = ast.Node{
            .string = .{ .val = tok.literal },
        };

        self.program.append(string) catch @panic("failed to append node");
        self.indexes.append(@intCast(self.program.items.len - 1)) catch @panic("failed to append index");
        return null;
    }

    fn parseBoolean(self: *Parser) ?ParseErr {
        const tok = self.cur;
        self.advance();
        const boolean = ast.Node{
            .boolean = .{ .val = std.mem.eql(u8, tok.literal, "true") },
        };

        self.program.append(boolean) catch @panic("failed to append node");
        self.indexes.append(@intCast(self.program.items.len - 1)) catch @panic("failed to append index");
        return null;
    }

    fn parseBinOp(self: *Parser) ?ParseErr {
        const tok = self.cur;
        if (self.program.items.len < 2) {
            return ParseErr{
                .type = ParserErrType.NotEnoughArguments,
                .row = tok.row,
                .col = tok.col,
            };
        }

        const rhs_index = self.indexes.pop();
        const lhs_index = self.indexes.pop();

        const binop_type = switch (tok.type) {
            .Plus => ast.BinOpType.Add,
            .Minus => ast.BinOpType.Sub,
            .Asterisk => ast.BinOpType.Multiply,
            else => @panic("invalid binop"),
        };

        self.advance();

        const binop = ast.Node{
            .binop = .{ .op = binop_type, .lhs = lhs_index, .rhs = rhs_index },
        };

        self.program.append(binop) catch @panic("failed to append node");
        self.indexes.append(@intCast(self.program.items.len - 1)) catch @panic("failed to append index");
        return null;
    }

    fn parseTable(self: *Parser) ?ParseErr {
        assert(self.cur.type == .LSquiggly);
        self.advance();
        const new_indices = std.ArrayList(u32).init(self.allocator);
        var newParser = Parser{
            .allocator = self.allocator,
            .tokens = self.tokens,
            .program = self.program,
            .indexes = new_indices,
            .cur = self.cur,
            .read_pos = self.read_pos,
        };

        std.debug.print("parsing table {any}\n", .{newParser.tokens[self.read_pos..]});

        // const table = ast.Node{ .table = .{ .items = []u32{} } };

        var index: usize = 0;
        const upper_bound = self.tokens.len - self.read_pos;
        while (self.cur.type != .EOF) : (index += 1) {
            std.debug.print("iteration: {}\n", .{index});
            if (index >= upper_bound) {
                @panic("reached table item upper bound");
            }
            if (newParser.parseExpression()) |err| {
                if (err.type != ParserErrType.InvalidToken) {
                    return err;
                }
                const tok = newParser.cur;
                if (tok.type == .Comma) {
                    std.debug.print("found our item, {any}\n", .{newParser.indexes.items});
                    newParser.advance();
                }
                if (tok.type == .RSquiggly) {
                    std.debug.print("found our table\n", .{});
                    newParser.advance();
                }
                std.debug.print("error invalid token '{any}' at {}:{}\n", .{ tok, err.row, err.col });
            }
        }
        return null;
    }

    fn parseExpression(self: *Parser) ?ParseErr {
        switch (self.cur.type) {
            .Number => return self.parseNumber(),
            .String => return self.parseString(),
            .Boolean => return self.parseBoolean(),
            .Plus,
            .Minus,
            .Asterisk,
            .Slash,
            .LessThan,
            .LessThanEqual,
            .GreaterThan,
            .GreaterThanEqual,
            .Equal,
            .NotEqual,
            .And,
            .Or,
            => return self.parseBinOp(),
            .LSquiggly => return self.parseTable(),
            else => return ParseErr{
                .type = ParserErrType.InvalidToken,
                .row = self.cur.row,
                .col = self.cur.col,
            },
        }

        // t := p.currToken
        //
        // switch t.Type {
        // case token.LSquiggly:
        // 	node, err := p.parseTable()
        // 	return node, err
        //
        // case token.LParen:
        // 	node, err := p.parseFn()
        // 	return node, err
        //
        // case token.Ident:
        // 	node, err := p.parseIdent()
        // 	return node, err
        //
        // default:
        // 	return nil, fmt.Errorf("%w: token '%s' at pos '%d:%d'", ErrInvalidToken, t.Literal, t.Row, t.Col)
        // }
    }

    pub fn parse(self: *Parser) struct { program: []ast.Node, indexes: []u32 } {
        var index: usize = 0;
        const upper_bound = self.tokens.len;
        while (self.cur.type != .EOF) : (index += 1) {
            if (index >= upper_bound) {
                @panic("reached parser's upper bound");
            }
            if (self.parseExpression()) |err| {
                std.debug.print("error at: {}:{}\n", .{ err.row, err.col });
                @panic("Parsing error");
            }
        }

        return .{ .program = self.program.items, .indexes = self.indexes.items };
    }
};

test "advance" {
    const allocator = std.testing.allocator;
    const tokens = [_]token.Token{
        .{ .type = .Number, .row = 0, .col = 0, .literal = "0" },
        .{ .type = .Number, .row = 0, .col = 2, .literal = "123" },
        .{ .type = .EOF, .row = 0, .col = 4, .literal = "" },
    };

    var parser = Parser.init(allocator, &tokens);
    defer parser.deinit();

    try std.testing.expectEqualDeep(tokens[0], parser.cur);
    parser.advance();
    try std.testing.expectEqualDeep(tokens[1], parser.cur);
    parser.advance();
    try std.testing.expectEqualDeep(tokens[2], parser.cur);
    parser.advance();
    try std.testing.expectEqualDeep(tokens[2], parser.cur);
}

test "parse_basics" {
    const allocator = std.testing.allocator;
    const tokens = [_]token.Token{
        .{ .type = .Number, .row = 0, .col = 0, .literal = "0" },
        .{ .type = .Number, .row = 0, .col = 2, .literal = "123" },
        .{ .type = .String, .row = 0, .col = 6, .literal = "hello there" },
        .{ .type = .Boolean, .row = 0, .col = 18, .literal = "true" },
        .{ .type = .Boolean, .row = 0, .col = 23, .literal = "false" },
        .{ .type = .EOF, .row = 0, .col = 27, .literal = "" },
    };

    const expected_ast = [_]ast.Node{
        ast.Node{ .number = .{ .val = 0 } },
        ast.Node{ .number = .{ .val = 123 } },
        ast.Node{ .string = .{ .val = "hello there" } },
        ast.Node{ .boolean = .{ .val = true } },
        ast.Node{ .boolean = .{ .val = false } },
    };

    const expected_ast_indexes = [_]u32{ 0, 1, 2, 3, 4 };

    var parser = Parser.init(allocator, &tokens);
    defer parser.deinit();

    const parsed_ast = parser.parse();
    try std.testing.expectEqual(expected_ast.len, parsed_ast.program.len);
    try std.testing.expectEqual(expected_ast_indexes.len, parsed_ast.indexes.len);

    try std.testing.expectEqualSlices(u32, &expected_ast_indexes, parsed_ast.indexes);

    var i: usize = 0;
    while (i < expected_ast.len) : (i += 1) {
        const expected = expected_ast[i].toString(allocator) catch @panic("failed to stringify ast");
        defer allocator.free(expected);

        const got = parsed_ast.program[i].toString(allocator) catch @panic("failed to stringify ast");
        defer allocator.free(got);

        try std.testing.expectEqualStrings(expected, got);
    }
}

test "parse_table" {
    const allocator = std.testing.allocator;
    const tokens = [_]token.Token{
        .{ .type = .LSquiggly, .row = 0, .col = 0, .literal = "{" },
        .{ .type = .Number, .row = 0, .col = 1, .literal = "1" },
        .{ .type = .Comma, .row = 0, .col = 2, .literal = "," },
        .{ .type = .Number, .row = 0, .col = 3, .literal = "2" },
        .{ .type = .RSquiggly, .row = 0, .col = 4, .literal = "}" },
        .{ .type = .EOF, .row = 0, .col = 5, .literal = "" },
    };

    const expected_ast = [_]ast.Node{
        ast.Node{
            .table = .{
                .items = &[_]ast.Node{
                    ast.Node{ .number = .{ .val = 1 } },
                    ast.Node{ .number = .{ .val = 2 } },
                },
            },
        },
    };

    const expected_ast_indexes = [_]u32{0};

    var parser = Parser.init(allocator, &tokens);
    defer parser.deinit();

    const parsed_ast = parser.parse();
    try std.testing.expectEqual(expected_ast.len, parsed_ast.program.len);
    try std.testing.expectEqual(expected_ast_indexes.len, parsed_ast.indexes.len);

    try std.testing.expectEqualSlices(u32, &expected_ast_indexes, parsed_ast.indexes);

    var i: usize = 0;
    while (i < expected_ast.len) : (i += 1) {
        const expected = expected_ast[i].toString(allocator) catch @panic("failed to stringify ast");
        defer allocator.free(expected);

        const got = parsed_ast.program[i].toString(allocator) catch @panic("failed to stringify ast");
        defer allocator.free(got);

        try std.testing.expectEqualStrings(expected, got);
    }
}

test "parse_advanced" {
    const allocator = std.testing.allocator;
    const tokens = [_]token.Token{
        .{ .type = .Number, .row = 0, .col = 0, .literal = "1" },
        .{ .type = .Number, .row = 0, .col = 2, .literal = "2" },
        .{ .type = .Plus, .row = 0, .col = 4, .literal = "+" },
        .{ .type = .Number, .row = 0, .col = 6, .literal = "3" },
        .{ .type = .Asterisk, .row = 0, .col = 8, .literal = "*" },
        .{ .type = .EOF, .row = 0, .col = 4, .literal = "" },
    };

    const expected_ast = [_]ast.Node{
        ast.Node{ .number = .{ .val = 1 } },
        ast.Node{ .number = .{ .val = 2 } },
        ast.Node{ .binop = .{ .op = .Add, .lhs = 0, .rhs = 1 } },
        ast.Node{ .number = .{ .val = 3 } },
        ast.Node{ .binop = .{ .op = .Multiply, .lhs = 2, .rhs = 3 } },
    };

    const expected_ast_indexes = [_]u32{4};

    var parser = Parser.init(allocator, &tokens);
    defer parser.deinit();

    const parsed_ast = parser.parse();
    try std.testing.expectEqual(expected_ast.len, parsed_ast.program.len);
    try std.testing.expectEqual(expected_ast_indexes.len, parsed_ast.indexes.len);

    try std.testing.expectEqualSlices(u32, &expected_ast_indexes, parsed_ast.indexes);

    var i: usize = 0;
    while (i < expected_ast.len) : (i += 1) {
        const expected = expected_ast[i].toString(allocator) catch @panic("failed to stringify ast");
        defer allocator.free(expected);

        const got = parsed_ast.program[i].toString(allocator) catch @panic("failed to stringify ast");
        defer allocator.free(got);

        try std.testing.expectEqualStrings(expected, got);
    }
}
